Reorganize Security Best Practices into Better Categories (#3034)
dfinity/portal: docs/developer-docs/security/security-best-practices/misc.mdx

---
sidebar_position: 11

> sidebar_label: Miscellaneous

Author: Done

keywords: [security, concept]
---

import { MarkdownChipRow } from "/src/components/Chip/MarkdownChipRow";

# Security best practices: Miscellaneous

<MarkdownChipRow labels={["Intermediate", "Security", "Concept"]} />

> ## Data confidentiality on the Internet Computer

Author: Done

### Security concern

> When storing data on the Internet Computer, there are two levels of data access.

Reviewer: ```suggestion
When storing data on ICP, there are two levels of data access.
```

Author: Done

In particular, a canister can ensure that responses to ingress messages do not leak data to external users, other than the sender, by including high-entropy data in the response. Or, a canister can ensure that data in certified variables is not leaked by adding high-entropy data to the variables that should be kept confidential.

Additionally, similarly to ingress message responses, a canister's private custom sections that contain low-entropy data could leak to unauthorized users. Therefore, a sufficent level of entropy for canister private custom sections should be used. 128 bits is recommended. If the data does not have enough entropy itself, then adding some artificial data using randomness would be recommended.

## Using secure randomness in canisters

Canister developers often require access to secure randomness in their canisters to perform certain operations. The requirements for a secure randomness source include
* **Unbiased:** The value shouldn't be influenced by anyone.
* **Unpredictable:** The value is unknown to anyone before it is generated.

> The Internet Computer exposes the system API [`raw_rand`](/docs/current/references/ic-interface-spec#ic-raw_rand) for this exact purpose, which accepts no input and returns 32 bytes of cryptographically secure randomness. The secure randomness guarantees of `raw_rand` are achieved by the cryptographic properties of the [RandomTape](https://internetcomputer.org/how-it-works/execution-layer/#random-number-generation). It is always recommended to use `raw_rand` as a source of randomness in canisters and **avoid** using other sources with low entropy such as current time.

Reviewer: ```suggestion
ICP exposes the system API [`raw_rand`](/docs/current/references/ic-interface-spec#ic-raw_rand) for this exact purpose, which accepts no input and returns 32 bytes of cryptographically secure randomness. The secure randomness guarantees of `raw_rand` are achieved by the cryptographic properties of the [RandomTape](https://internetcomputer.org/how-it-works/execution-layer/#random-number-generation). It is always recommended to use `raw_rand` as a source of randomness in canisters and **avoid** using other sources with low entropy such as current time.
```

Author: Done

> To illustrate the usage of `raw_rand`, we discuss two examples in Motoko and Rust below, while weighing the benefits and caveats around using them.

Reviewer: ```suggestion
To illustrate the usage of `raw_rand`, two examples in Motoko and Rust can be found below including the benefits and caveats around using them.
```

Author: Done

        ic_cdk::spawn(async {
            seed_randomness().await;
            // Schedule reseeding on a timer with duration SEEDING_INTERVAL
            schedule_seeding(SEEDING_INTERVAL);
        })
    });
}

#[ic_cdk::init]
fn init() {
    // Initialize randomness during canister install or reinstall
    schedule_seeding(Duration::ZERO);
}

#[ic_cdk::post_upgrade]
fn post_upgrade() {
    // Initialize randomness after a canister upgrade
    schedule_seeding(Duration::ZERO);
}

// This must always be an update method or the PRNG state won't be updated
#[ic_cdk::update]
fn random_bytes(size: u32) -> RandomBytesResult {
    let mut buf = vec![0; size as usize];
    RNG.with_borrow_mut(|rng| match rng.as_mut() {
        Some(rand) => {
            rand.fill_bytes(&mut buf);
            Ok(hex::encode(buf))
        }
        None => Err(RngError::RngNotInitialized(
            "Randomness is not initialized. Please try again later".to_string(),
        )),
    })
}
```

#### Benefits:
- This method scales for large random bytes as `raw_rand` needs be called only once and subsequent PRNG computation is local to the canister.

#### Caveats:

> - The `setup_randomness` must **always** be initialized in both the `init` and `post_upgrade` hook as `init` is not invoked during an canister upgrade [see here](/docs/current/references/ic-interface-spec#system-api-upgrades).

Author: Done

- The `init` and `post_upgrade` methods doesn't allow async calls and thus a timer is immediately scheduled to seed the randomness.

> - Once the seed is initialized, the outcome of all future `random_bytes` is predicable to anyone having the seed (node providers) as the PRNG is deterministic. This breaks the unpredicatable property of secure randomness. Hence, to balance security vs performance, we recommend frequently reseeding the PRNG on a timer. The example above already does this with duration of **1 hour**. However, based on the sensitivity of their dApp, developers can choose an appropriate reseeding interval by setting `SEEDING_INTERVAL`.

Reviewer: ```suggestion
- Once the seed is initialized, the outcome of all future `random_bytes` is predicable to anyone having the seed (node providers) as the PRNG is deterministic. This breaks the unpredicatable property of secure randomness. Hence, to balance security vs performance, we recommend frequently reseeding the PRNG on a timer. The example above already does this with duration of **1 hour**. However, based on the sensitivity of their dapp, developers can choose an appropriate reseeding interval by setting `SEEDING_INTERVAL`.
```

Author: Done

- The `random_bytes` must **always** be an `update` method, so the PRNG can preserve the state and offer unique randomness on every request.

## Test your canister code even in presence of system API calls

### Security concern

Since canisters interact with the system API, it is harder to test the code because unit tests cannot call the system API. This may lead to lack of unit tests.

### Recommendation

- Create loosely coupled modules that do not depend on the system API and unit test those. See this [recommendation](https://mmapped.blog/posts/01-effective-rust-canisters.html#target-independent) (from [effective Rust canisters](https://mmapped.blog/posts/01-effective-rust-canisters.html)).

> - For the parts that still interact with the system API: create a thin abstraction of the System API that is faked in unit tests. See the [recommendation](https://mmapped.blog/posts/01-effective-rust-canisters.html#target-independent) (from [effective Rust canisters](https://mmapped.blog/posts/01-effective-rust-canisters.html)). For example, one can implement a ‘Runtime’ as follows and then use the ‘MockRuntime’ in tests (code by Dimitris Sarlis):

Author: Done

> <!-- -->

Author: Done

    (...)
}

#[async_trait]
impl Runtime for MockRuntime {
    fn caller(&self) -> Result<Principal, String> {
        Ok(self.caller)
    }

    fn id(&self) -> Principal {
        self.canister_id
    }

    fn time(&self) -> u64 {
        self.time
    }

    (...)

}
```

## Make canister builds reproducible

### Security concern

It should be possible to verify that a canister does what it claims to do. ICP provides a SHA256 hash of the deployed WASM module. In order for this to be useful, the canister build has to be reproducible.

### Recommendation

Make canister builds reproducible. See this [recommendation](https://mmapped.blog/posts/01-effective-rust-canisters.html#reproducible-builds) (from [effective Rust canisters](https://mmapped.blog/posts/01-effective-rust-canisters.html)). See also the [developer docs on this](/docs/current/developer-docs/smart-contracts/test/reproducible-builds).

## Don’t rely on time being strictly monotonic

### Security concern

The time read from the System API is monotonic, but not strictly monotonic. Thus, two subsequent calls can return the same time, which could lead to security bugs when the time API is used.

### Recommendation

> See the "Time is not strictly monotonic" section in [how to audit an Internet Computer canister](https://www.joachim-breitner.de/blog/788-How_to_audit_an_Internet_Computer_canister).

Reviewer: ```suggestion
See the "Time is not strictly monotonic" section in [how to audit an ICP canister](https://www.joachim-breitner.de/blog/788-How_to_audit_an_Internet_Computer_canister).
```

Author: Done
